Lib/test/test_threading.py

```diff
@@ -482,6 +482,47 @@ def test_enumerate_after_join(self):
         finally:
             sys.setswitchinterval(old_interval)
 
+    def test_join_from_multiple_threads(self):
+        # Thread.join() should be thread-safe
+        errors = []
+
+        def worker():
+            time.sleep(0.005)
+
+        def joiner(thread):
+            try:
+                thread.join()
+            except Exception as e:
+                errors.append(e)
+
+        for N in range(2, 20):
+            threads = [threading.Thread(target=worker)]
+            for i in range(N):
+                threads.append(threading.Thread(target=joiner,
+                                                args=(threads[0],)))
+            for t in threads:
+                t.start()
+            time.sleep(0.01)
+            for t in threads:
+                t.join()
+            if errors:
+                raise errors[0]
```

Comment:
(feel free to ignore this)

You could also raise an `ExceptionGroup`:

```suggestion
                raise ExceptionGroup('Thread.join() failures', errors)
```

However, for this test the important fact is that a failure happened at all, so using `ExceptionGroup` doesn't add much value.